include/dali/core/mm/malloc_resource.h

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_MM_MALLOC_RESOURCE_H_
+#define DALI_CORE_MM_MALLOC_RESOURCE_H_
+
+#include <stdlib.h>
+#include <malloc.h>
+#include "dali/core/mm/memory_resource.h"
+
+namespace dali {
+namespace mm {
+
+/**
+ * @brief A memory resource that manages host memory with std::aligned_alloc and std::free
+ */
+class malloc_memory_resource : public memory_resource {
+  void *do_allocate(size_t bytes, size_t alignment) override {
+    return memalign(alignment, bytes + sizeof(int));
+  }
+
+  void do_deallocate(void *ptr, size_t bytes, size_t alignment) override {
+    return free(ptr);
+  }
+
+  bool do_is_equal(const memory_resource &other) const noexcept override {
+    return dynamic_cast<const malloc_memory_resource*>(&other) != nullptr;
+  }
+
+ public:
+  static malloc_memory_resource &instance() {
+    static malloc_memory_resource inst;
+    return inst;
+  }
```

Comment:
It can still be inherited from and maybe the derived class will not be a singleton. Anyway, the singleton part is here only to provide a usable address to `upstream` parameters without caring too much for the object lifetime. For other purposes, one can still create an instance.